docs/ocis/adr/0020-file-search-query-language.md

```diff
@@ -0,0 +1,104 @@
+
+* [e.g., compromising quality attribute, follow-up decisions required, …]
+* …
+
+## Pros and Cons of the Options
+
+### Keyword Query Language (KQL)
+
+Implement a search based on the Keyword Query Language (KQL), adopted from Sharepoint.
+
+* Good, because microsoft already uses it togethe with the GraphAPI
+* Bad, because there is no go package
+
+### Simplified Query
+
+Implement a very simple search approach: Return all files which contain at least one of the keywords in their name, path, alias or selected metadata.
+
+* Good, because that covers 80% of the users needs
+* Good, because it is very straightforward
+* Bad, because it is below the industry standard
+* Bad, because it only provides one search query
+
+### Lucene Query Language
+
+The Lucene Query Parser syntax supports advanced queries like term, phrase, wildcard, fuzzy search, proximity search, regular expressions, boosting, boolean operators and grouping. It is a well known query syntax used by the Apache Lucene Project. Popular Platforms like Wikipedia are using Lucene or Solr, which is the successor of Lucene
+
+* Good, because it is a well documented and powerful syntax
+* Good, because it is very close to the Elasticsearch and the Solr syntax which enhances compatibility
+* Bad, because there is no powerful and well tested query parser for golang available
+* Bad, because it adds complexity and fulfilling all the different query usecases can be an "uphill battle"
+
+### Solr Query Language
+
+Solr is highly reliable, scalable and fault tolerant, providing distributed indexing, replication and load-balanced querying, automated failover and recovery, centralized configuration and more. Solr powers the search and navigation features of many of the world's largest internet sites.
+
+* Good, because it is a well documented and powerful syntax
+* Good, because it is very close to the Elasticsearch and the Lucene syntax which enhances compatibility
+* Good, because it has a strong community with large resources and knowledge
+* Bad, because it adds complexity and fulfilling all the different query usecases can be an "uphill battle"
+
+### Elasticsearch Query Language
+
+Elasticsearch provides a full Query DSL (Domain Specific Language) based on JSON to define queries. Think of the Query DSL as an AST (Abstract Syntax Tree) of queries, consisting of two types of clauses. It is able to combine multiple query types into compound queries. It is also a successor of Solr.
+
+* Good, because it is a well documented and powerful syntax
+* Good, because it is very close to the Elasticsearch and the Solr syntax which enhances compatibility
+* Good, because there is a stable and well tested go client which brings a query builder
+* Good, because it could be used as the query language which supports different search backends by just implementing what is needed for our usecase
+* Bad, because it adds complexity and fulfilling all the different query usecases can be an "uphill battle"
+
```

Comment:
I am not sure why we discuss these Query Languages. We agree, that we do not want to expose these QL to the users, at least in the first iteration. So it is part of the search services, if it once supports multiple search backends, to translate the "simple approach" to the backend QL to get the best results.

Reply:
We need that query language internally.